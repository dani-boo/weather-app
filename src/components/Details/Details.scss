.wrapper {
  display: flex;
  flex-direction: column;
  min-height: 100px;
}

.warning {
  color: red;
}

.long-list {
  display: flex;
  flex-wrap: wrap;
  margin-bottom: 100px;
  @media only screen and (max-width: 414px) {
    padding-bottom: 64px;
  }
}

.day {
  flex: 1 0 21%;
  padding: 12px;
}

.day:hover {
  box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);
}

.image {
  width: 5em;
  height: auto;
}

dd {
  margin: auto;
}

.city-name {
  font-size: 2em;
}
