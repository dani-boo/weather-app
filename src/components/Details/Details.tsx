import React, { useState } from 'react'
import Temperature from '../Temperature/Temperature'
import handleSetFilter from '../../utils/handleSetState'
import './Details.scss'

interface dataObj {
  city_name?: string
  datetime: string
  max_temp: number
  min_temp: number
  temp: number
  weather: {
    icon: string
  }
}

type dataArr = dataObj[]

export interface weatherData {
  data?: dataArr
  city_name?: string
}
export interface responseData {
  weatherObj?: weatherData
  error?: boolean
  type?: string
  filter?: boolean
}

export const filterData = (props: responseData) => (
  minOrMax: string,
  inputValue: number
) => (setData: (value: any) => void) => {
  const filtered = props?.weatherObj?.data?.filter((day) => {
    if (minOrMax === 'min') return day.min_temp <= inputValue
    if (minOrMax === 'max') return day.max_temp >= inputValue
  })
  setData(filtered)
}

export const handleNumberInput = (setValue: (value: number) => void) => (
  e: React.FormEvent<HTMLInputElement>
) => {
  const target = e.target as HTMLInputElement
  setValue(parseInt(target.value))
}

export const getFiltered = (
  setRenderFiltered: (value: boolean) => void,
  props: responseData,
  filter: string,
  value: number,
  setData: (value: any) => void
) => (e: React.SyntheticEvent) => {
  e.preventDefault()
  filterData(props)(filter, value)(setData)
  setRenderFiltered(true)
}

export const getDataToRender = (
  renderFiltered: boolean,
  dataArr: dataArr | undefined,
  props: responseData
) => {
  return renderFiltered ? dataArr : props.weatherObj?.data
}

export default function Details(props: responseData) {
  const [value, setValue] = useState<number>(0)
  const [filter, setFilter] = useState<string>('')
  const [data, setData] = useState<[]>([])
  const [renderFiltered, setRenderFiltered] = useState(false)

  const selectFilter = handleSetFilter(setFilter)

  const handleInput = handleNumberInput(setValue)

  const getFilteredTemps = getFiltered(
    setRenderFiltered,
    props,
    filter,
    value,
    setData
  )

  const dataToRender = getDataToRender(renderFiltered, data, props)

  return (
    <div className='wrapper'>
      {props.error && <p className='warning'>Please select a city.</p>}
      {props.weatherObj?.data && props.type === 'current' && (
        <div>
          <h3 className='city-name'>
            <strong>{props.weatherObj.data[0].city_name}</strong>
          </h3>
          <p>{Math.round(props.weatherObj.data[0].temp)} degrees celsius</p>
          <img
            alt='weather icon'
            src={`https://www.weatherbit.io/static/img/icons/${props.weatherObj.data[0].weather.icon}.png`}
          />
        </div>
      )}
      {props.weatherObj?.data && props.type === 'sixteenDays' && (
        <>
          <Temperature
            handleNumberInput={handleInput}
            selectFilter={selectFilter}
            getTemperatures={getFilteredTemps}
          />
          <h3 className='city-name'>
            <strong>{props.weatherObj.city_name}</strong>
          </h3>
          <div className='long-list'>
            {dataToRender !== undefined &&
              dataToRender.map((day) => (
                <dl className='day' key={day.datetime}>
                  <dt>{day.datetime}</dt>
                  <img
                    alt='weather icon'
                    className='image'
                    src={`https://www.weatherbit.io/static/img/icons/${day.weather.icon}.png`}
                  />
                  <dd>
                    {Math.round(day.max_temp)}°C {Math.round(day.min_temp)}°C
                  </dd>
                </dl>
              ))}
          </div>
        </>
      )}
    </div>
  )
}
