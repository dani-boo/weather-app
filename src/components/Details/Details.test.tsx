import React from 'react'
import { shallow } from 'enzyme'
import set from 'lodash/set'
import Details, {
  filterData,
  handleNumberInput,
  getDataToRender,
  getFiltered,
} from './Details'

const obj = {
  weatherObj: {
    data: [
      {
        datetime: 'datetime',
        min_temp: 4,
        max_temp: 30,
        temp: 0,
        weather: { icon: 'icon' },
      },
      {
        datetime: 'datetime',
        min_temp: 3,
        max_temp: 20,
        temp: 0,
        weather: { icon: 'icon' },
      },
      {
        datetime: 'datetime',
        min_temp: 20,
        max_temp: 10,
        temp: 0,
        weather: { icon: 'icon' },
      },
      {
        datetime: 'datetime',
        min_temp: 2,
        max_temp: 50,
        temp: 0,
        weather: { icon: 'icon' },
      },
      {
        datetime: 'datetime',
        min_temp: 90,
        max_temp: 45,
        temp: 0,
        weather: { icon: 'icon' },
      },
      {
        datetime: 'datetime',
        min_temp: 0,
        max_temp: 5,
        temp: 0,
        weather: { icon: 'icon' },
      },
      {
        datetime: 'datetime',
        min_temp: 79,
        max_temp: 82,
        temp: 0,
        weather: { icon: 'icon' },
      },
    ],
  },
}

const filteredMin = [
  {
    datetime: 'datetime',
    min_temp: 4,
    max_temp: 30,
    temp: 0,
    weather: { icon: 'icon' },
  },
  {
    datetime: 'datetime',
    min_temp: 3,
    max_temp: 20,
    temp: 0,
    weather: { icon: 'icon' },
  },
  {
    datetime: 'datetime',
    min_temp: 2,
    max_temp: 50,
    temp: 0,
    weather: { icon: 'icon' },
  },
  {
    datetime: 'datetime',
    min_temp: 0,
    max_temp: 5,
    temp: 0,
    weather: { icon: 'icon' },
  },
]

describe('filterData', () => {
  const setData = jest.fn()
  it('filters data min_temps and calls setData with filtered results', () => {
    filterData(obj)('min', 15)(setData)
    expect(setData).toHaveBeenCalledWith(filteredMin)
  })
  it('filters data max_temps and calls setData with filtered results', () => {
    const filtered = [
      {
        datetime: 'datetime',
        min_temp: 79,
        max_temp: 82,
        temp: 0,
        weather: { icon: 'icon' },
      },
    ]

    filterData(obj)('max', 79)(setData)
    expect(setData).toHaveBeenCalledWith(filtered)
  })
  it("doesn't filter if filters are not min or max", () => {
    filterData(obj)('fake filter', 50)(setData)
    expect(setData).toHaveBeenCalledWith([])
  })
})

describe('handleNumberInput', () => {
  it('should call setValue() with selected value', () => {
    const setValue = jest.fn()
    const event = {
      target: { value: '20' },
    } as React.ChangeEvent<HTMLInputElement>

    handleNumberInput(setValue)(event)
    expect(setValue).toHaveBeenCalledWith(20)
  })
})

describe('getFiltered', () => {
  it('calls filter functions and sets filtered to true', () => {
    const setRenderFiltered = jest.fn()
    const e = ({
      preventDefault: jest.fn(),
    } as unknown) as React.ChangeEvent<HTMLInputElement>
    const setData = jest.fn()
    getFiltered(setRenderFiltered, obj, 'min', 15, setData)(e)
    expect(e.preventDefault).toHaveBeenCalled()
    expect(setRenderFiltered).toHaveBeenCalledWith(true)
  })
})

describe('getDataToRender', () => {
  it('should return data array if renderFiltered is true', () => {
    const filtered = getDataToRender(true, filteredMin, obj)
    expect(filtered).toBe(filteredMin)
  })
  it('should return weather object if renderFiltered is false', () => {
    const filtered = getDataToRender(false, filteredMin, obj)
    expect(filtered).toBe(obj.weatherObj?.data)
  })
})

describe('Details', () => {
  it('should render with error', () => {
    set(obj, 'error', true)
    const wrapper = shallow(<Details {...obj} />)
    expect(wrapper.text().includes('Please select a city.')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should render current data', () => {
    set(obj, 'error', false)
    set(obj, 'type', 'current')
    const wrapper = shallow(<Details {...obj} />)
    expect(wrapper.text().includes('degrees celsius')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should render 16 days data', () => {
    set(obj, 'type', 'sixteenDays')
    const wrapper = shallow(<Details {...obj} />)
    expect(wrapper.text().includes('Â°C')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
