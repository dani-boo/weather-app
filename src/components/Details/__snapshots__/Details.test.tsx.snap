// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details should render 16 days data 1`] = `"<div class=\\"wrapper\\"><h3>Search by temperature:</h3><form><label class=\\"field\\"><input type=\\"number\\" name=\\"temperature\\" placeholder=\\"\\"/></label><label><input type=\\"radio\\" name=\\"filter\\" value=\\"min\\"/>min</label><label><input type=\\"radio\\" name=\\"filter\\" value=\\"max\\"/>max</label><button class=\\"button\\" type=\\"submit\\">Filter results</button></form><button class=\\"button\\">Refresh page</button><h3 class=\\"city-name\\"><strong></strong></h3><div class=\\"long-list\\"><dl class=\\"day\\"><dt>datetime</dt><img alt=\\"weather icon\\" class=\\"image\\" src=\\"https://www.weatherbit.io/static/img/icons/icon.png\\"/><dd>30°C 4°C</dd></dl><dl class=\\"day\\"><dt>datetime</dt><img alt=\\"weather icon\\" class=\\"image\\" src=\\"https://www.weatherbit.io/static/img/icons/icon.png\\"/><dd>20°C 3°C</dd></dl><dl class=\\"day\\"><dt>datetime</dt><img alt=\\"weather icon\\" class=\\"image\\" src=\\"https://www.weatherbit.io/static/img/icons/icon.png\\"/><dd>10°C 20°C</dd></dl><dl class=\\"day\\"><dt>datetime</dt><img alt=\\"weather icon\\" class=\\"image\\" src=\\"https://www.weatherbit.io/static/img/icons/icon.png\\"/><dd>50°C 2°C</dd></dl><dl class=\\"day\\"><dt>datetime</dt><img alt=\\"weather icon\\" class=\\"image\\" src=\\"https://www.weatherbit.io/static/img/icons/icon.png\\"/><dd>45°C 90°C</dd></dl><dl class=\\"day\\"><dt>datetime</dt><img alt=\\"weather icon\\" class=\\"image\\" src=\\"https://www.weatherbit.io/static/img/icons/icon.png\\"/><dd>5°C 0°C</dd></dl><dl class=\\"day\\"><dt>datetime</dt><img alt=\\"weather icon\\" class=\\"image\\" src=\\"https://www.weatherbit.io/static/img/icons/icon.png\\"/><dd>82°C 79°C</dd></dl></div></div>"`;

exports[`Details should render current data 1`] = `"<div class=\\"wrapper\\"><div><h3 class=\\"city-name\\"><strong></strong></h3><p>0 degrees celsius</p><img alt=\\"weather icon\\" src=\\"https://www.weatherbit.io/static/img/icons/icon.png\\"/></div></div>"`;

exports[`Details should render with error 1`] = `"<div class=\\"wrapper\\"><p class=\\"warning\\">Please select a city.</p></div>"`;
